README.md
# This project is still under development.

This is a small game meant for the purpose of developing and testing algorithms/AI to solve this game.

The game is based on a game by Lars Feßen. Check the game out [here](https://play.google.com/store/apps/details?id=com.kila.zahlenspiel2.lars).

---

## Installation
Install [Node.js](https://nodejs.org/en/) 
and [Yarn](https://classic.yarnpkg.com/en/docs/install#windows-stable)

`npm install --global yarn`

Run
`yarn`
`yarn run dev`

---

The project is created with [Phaser 3.](https://github.com/photonstorm/phaser)
This project is licensed under the MIT License - see the LICENSE.md file for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [0.3.0] - 2022-12-07

### Added

- help function

## [0.2.0] - 2022-12-07

### Added

- menu
- findNeighbour function (currently unused -> future use for neighbour highlighting)
- check function
- undo function (last cross)
- restart function

### Changed

- Board creation through constructor
- Cross function checks for crossed elements being neighbours and values being equal or 10

## [0.1.0] - 2022-12-05

### Added

- Board creation
- Crossing numbers (without checks)

src/config.ts
import { GameScene } from './scenes/game-scene';
import { StartScene } from './scenes/start-scene';

export const Constants: {
  height: number,
  width: number,
  menuHeight: number,
  gamemodes: {name: string, startarray: number[], length: number}[],
} = {
  height: 800,
  width: 450,
  menuHeight: 70,
  gamemodes: [
    {name: '10 random numbers', startarray: null, length: 10},
    {name: '20 random numbers', startarray: null, length: 20},
    {name: '1 - 12', startarray: [1,2,3,4,5,6,7,8,9,1,0,1,1,1,2], length: null},
    {name: '1 - 13', startarray: [1,2,3,4,5,6,7,8,9,1,0,1,1,1,2,1,3], length: null},
  ]
};

export const GameConfig: Phaser.Types.Core.GameConfig = {
  version: '1.0',
  type: Phaser.AUTO,
  scene: [StartScene, GameScene],
  scale: {
    parent: 'game',
    width: Constants.width,
    height: Constants.height
  },
  physics: {
    default: 'arcade',
    arcade: {
      gravity: { y: 0 },
    }
  },
  backgroundColor: 0x333333
};


src/scripts/gamecode.ts
export class GameCode {
    // 9xX
    private board: Array<Array<{visible: boolean, value: number}>>;
    private startnumbers: number[];
    private lastCross: {e1: {x: number, y: number}, e2: {x: number, y: number}};

    constructor(numbers: number[]) {
        this.board = [];
        this.startnumbers = numbers;
        this.appendArray([...this.startnumbers]);
    }

    public getBoardSizeY(): number {
        return this.board.length
    }

    public getBoardSizeX(index: number): number {
        return this.board[index].length;
    }

    public getValue(x: number, y: number): {visible: boolean, value: number} {
        return this.board[y][x];
    }

    public check(): void {
        this.lastCross = null;
        let numbers: number[] = [];
        for(let i = 0; i < this.board.length; i++) {
            for(let j = 0; j < this.board[i].length; j++) {
                if(this.board[i][j].visible) numbers.push(this.board[i][j].value);
            }
        }
        this.appendArray(numbers);
    }

    public undo(): void {
        if(!this.lastCross) return;
        this.board[this.lastCross.e1.y][this.lastCross.e1.x].visible = true;
        this.board[this.lastCross.e2.y][this.lastCross.e2.x].visible = true;
        this.lastCross = null;
    }

    public restart(): void {
        this.board = [];
        this.appendArray([...this.startnumbers]);
    }

    public findNeighbours(x: number, y: number): {
        right: {x: number, y: number}, 
        left: {x: number, y: number}, 
        top: {x: number, y: number}, 
        bottom: {x: number, y: number}
    } {
        return {
            right: this.rekRightNeighbour(x + 1, y), 
            left: this.rekLeftNeighbour(x - 1, y), 
            top: this.rekTopNeighbour(x, y - 1), 
            bottom: this.rekBottomNeighbour(x, y + 1)
        };
    }

    public help(): {e1: {x: number, y: number}, e2: {x: number, y: number}} {
        for(let i = 0; i < this.board.length; i++) {
            for(let j = 0; j < this.board[i].length; j++) {
                if(!this.board[i][j].visible) continue;
                let neighbor = this.rekBottomNeighbour(j, i + 1);
                if(neighbor) {
                    if(
                        this.board[i][j].value === this.board[neighbor.y][neighbor.x].value
                        || this.board[i][j].value + this.board[neighbor.y][neighbor.x].value == 10
                    ) return {e1: {x: j, y: i}, e2: neighbor};
                }
                neighbor = this.rekRightNeighbour(j + 1, i);
                if(neighbor) {
                    if(
                        this.board[i][j].value === this.board[neighbor.y][neighbor.x].value
                        || this.board[i][j].value + this.board[neighbor.y][neighbor.x].value == 10
                    ) return {e1: {x: j, y: i}, e2: neighbor};
                }
            }
        }
        return null;
    }

    public cross(e1: {x: number, y:number}, e2: {x: number, y:number}):void {
        if(e1.x === e2.x) {
            if(e1.y > e2.y) {
                let check = this.rekTopNeighbour(e1.x, e1.y - 1);
                if(check.x !== e2.x || check.y !== e2.y) return;
            } else {
                let check = this.rekBottomNeighbour(e1.x, e1.y + 1);
                if(check.x !== e2.x || check.y !== e2.y) return;
            }
        } else {
            if(e1.y > e2.y || (e1.x > e2.x && e1.y === e2.y)) {
                let check = this.rekLeftNeighbour(e1.x - 1, e1.y);
                if(check.x !== e2.x || check.y !== e2.y) return;
            } else {
                let check = this.rekRightNeighbour(e1.x + 1, e1.y);
                if(check.x !== e2.x || check.y !== e2.y) return;
            }
        }
        if(this.board[e1.y][e1.x].value !== this.board[e2.y][e2.x].value && this.board[e1.y][e1.x].value + this.board[e2.y][e2.x].value !== 10) return;
        this.board[e1.y][e1.x].visible = false;
        this.board[e2.y][e2.x].visible = false;
        this.lastCross = {e1: e1, e2: e2};
    }

    public clearRows(): void {
        this.lastCross = null;
        for(let i = 0; i < this.board.length - 1; i++) {
            let rDelete = true;
            for(let j = 0; j < this.board[i].length; j++) {
                if(this.board[i][j].visible) rDelete = false;
            }
            if(rDelete) {
                this.board.splice(i, 1);
                i--;
            }
        }
    }

    private appendArray(numbers: number[]):void {
        if(this.board.length !== 0) {
            if(this.board[this.board.length - 1].length < 9) {
                for(let i = this.board[this.board.length - 1].length - 1; i < 8; i++) {
                    if(numbers.length > 0) this.board[this.board.length - 1].push({visible: true, value: numbers.shift()});
                }
            }
        }
        while(numbers.length > 0) {
            this.board.push([]);
            for(let i = 0; i < 9; i++) {
                if(numbers.length > 0) this.board[this.board.length - 1].push({visible: true, value: numbers.shift()});
            }
        }
    }

    private rekLeftNeighbour(x: number, y: number): {x: number, y: number} {
        if (y < 0) return null;
        if (x < 0) return this.rekLeftNeighbour(8, y - 1);
        if (!this.board[y][x].visible) return this.rekLeftNeighbour(x - 1, y);
        return {x: x, y: y};
    }

    private rekRightNeighbour(x: number, y: number): {x: number, y: number} {
        if (y >= this.board.length) return null;
        if (y == this.board.length - 1 && x >= this.board[y].length) return null;
        if (x > 8) return this.rekRightNeighbour(0, y + 1);
        if (!this.board[y][x].visible) return this.rekRightNeighbour(x + 1, y);
        return {x: x, y: y};
    }

    private rekTopNeighbour(x: number, y: number): {x: number, y: number} {
        if (y < 0) return null;
        if (!this.board[y][x].visible) return this.rekTopNeighbour(x, y - 1);
        return {x: x, y: y};
    }

    private rekBottomNeighbour(x: number, y: number): {x: number, y: number} {
        if (y >= this.board.length) return null;
        if (y == this.board.length - 1 && x >= this.board[y].length) return null;
        if (!this.board[y][x].visible) return this.rekBottomNeighbour(x, y + 1);
        return {x: x, y: y};
    }
}

src/scenes/boot-scene.ts
export class BootScene extends Phaser.Scene {
    private loadingBar: Phaser.GameObjects.Graphics;
    private progressBar: Phaser.GameObjects.Graphics;
  
    constructor() {
      super({
        key: 'BootScene'
      });
    }
  
    preload(): void {
      // set the background and create loading bar
      //this.cameras.main.setBackgroundColor(0x98d687);
      this.createLoadingbar();
  
      // pass value to change the loading bar fill
      this.load.on(
        'progress',
        function (value: number) {
          this.progressBar.clear();
          this.progressBar.fillStyle(0xfff6d3, 1);
          this.progressBar.fillRect(
            this.cameras.main.width / 4,
            this.cameras.main.height / 2 - 16,
            (this.cameras.main.width / 2) * value,
            16
          );
        },
        this
      );
  
      // delete bar graphics, when loading complete
      this.load.on(
        'complete',
        function () {
          this.progressBar.destroy();
          this.loadingBar.destroy();
        },
        this
      );
  
      // load out package
      this.load.pack('preload', './assets/pack.json', 'preload');
    }
  
    update(): void {
      this.scene.start('GameScene');
    }
  
    private createLoadingbar(): void {
      this.loadingBar = this.add.graphics();
      this.loadingBar.fillStyle(0x5dae47, 1);
      this.loadingBar.fillRect(
        this.cameras.main.width / 4 - 2,
        this.cameras.main.height / 2 - 18,
        this.cameras.main.width / 2 + 4,
        20
      );
      this.progressBar = this.add.graphics();
    }
  }
  

src/scenes/start-scene.ts
import { Constants } from "../config";
import { randomInt, values } from "../scripts/shared";

export class StartScene extends Phaser.Scene {
    private menu: Array<{rect: Phaser.GameObjects.Rectangle, text: Phaser.GameObjects.Text}> = [];

    constructor() {
        super({
            key: "StartScene",
        });
    }

    create(): void {
        for(let i = 0; i < Constants.gamemodes.length; i++) {
            this.menu.push({
                rect: this.add.rectangle(20, 50 + 150 * i, Constants.width - 40, 100, 0x000000),
                text: this.add.text(30, 80 + 150 * i, Constants.gamemodes[i].name, { fontFamily: 'monospace', fontSize: '35px'}),
            });
            this.menu[i].rect.setOrigin(0,0);
            this.menu[i].rect.setInteractive();
            this.menu[i].rect.on('pointerup', () => this.startGame(i));
        }
    }

    private startGame(i: number): void {
        if(Constants.gamemodes[i].startarray) values.startarray = Constants.gamemodes[i].startarray;
        else {
            values.startarray = [];
            for(let j = 0; j < Constants.gamemodes[i].length; j++) {
                values.startarray.push(randomInt(1, 9));
            }
        }
        this.scene.stop('StartScene');
        this.scene.start('GameScene');
    }
}

src/scenes/game-scene.ts
import { GameCode } from "../scripts/gamecode";
import { Constants } from "../config";
import { values } from "../scripts/shared";

export class GameScene extends Phaser.Scene {
  private gameboard: GameCode;
  private gamestate: {
    selPos: {x: number, y: number},
    highlightedPos: {e1: {x: number, y: number}, e2: {x: number, y: number}},
    elements: Array<Array<{val: {value: number, visible: boolean}, rect: Phaser.GameObjects.Rectangle, number: Phaser.GameObjects.Text}>>,
    scrollHelper: any,
    cursors: any,
  } = {
    selPos: null,
    highlightedPos: null,
    elements: [],
    scrollHelper: null,
    cursors: null,
  };
  private menu: {
    bar: Phaser.GameObjects.Rectangle,
    buttons: {
      check: {rect: Phaser.GameObjects.Rectangle, text: Phaser.GameObjects.Text, symbol: Phaser.GameObjects.Sprite},
      undo: {rect: Phaser.GameObjects.Rectangle, text: Phaser.GameObjects.Text, symbol: Phaser.GameObjects.Sprite},
      help: {rect: Phaser.GameObjects.Rectangle, text: Phaser.GameObjects.Text, symbol: Phaser.GameObjects.Sprite},
      restart: {rect: Phaser.GameObjects.Rectangle, text: Phaser.GameObjects.Text, symbol: Phaser.GameObjects.Sprite},
      rowClear: {rect: Phaser.GameObjects.Rectangle, text: Phaser.GameObjects.Text, symbol: Phaser.GameObjects.Sprite},
    },
  } = {
    bar: null,
    buttons: {
      check: {rect: null, text: null, symbol: null},
      undo: {rect: null, text: null, symbol: null},
      help: {rect: null, text: null, symbol: null},
      restart: {rect: null, text: null, symbol: null},
      rowClear: {rect: null, text: null, symbol: null},
    },
  }

  constructor() {
    super({
      key: "GameScene",
    });
  }

  create(): void {
    this.gameboard = new GameCode(values.startarray);
    this.draw();
    this.createMenu();
    this.gamestate.scrollHelper = this.physics.add.sprite(0, Constants.height / 2, '').setScale(0);
    this.cameras.main.startFollow(this.gamestate.scrollHelper, true, 0.5, 0.5);
    this.gamestate.scrollHelper.setCollideWorldBounds(true);
    this.cameras.main.useBounds = true;
    this.gamestate.cursors = this.input.keyboard.createCursorKeys();
  }

  update(): void {
    if(this.gamestate.cursors.down.isDown) this.gamestate.scrollHelper.setVelocityY(200);
    else if (this.gamestate.cursors.up.isDown) this.gamestate.scrollHelper.setVelocityY(-200);
    else this.gamestate.scrollHelper.setVelocityY(0);
  }

  private selection(x: number, y: number): void {
    // future feature TODO's:
    // Custom Color Support
    // Neighbor Highlighting
    // Call draw function only if necessary
    if(!this.gamestate.elements[y][x].val.visible) return;
    if(!this.gamestate.selPos) {
      this.gamestate.selPos = {x: x, y: y};
      this.gamestate.elements[y][x].number.setColor('#0000ff');
      return;
    }
    if(this.gamestate.selPos.x == x && this.gamestate.selPos.y == y) {
      this.gamestate.selPos = null;
      this.gamestate.elements[y][x].number.setColor('#00ff00');
      return;
    }
    this.gameboard.cross(this.gamestate.selPos, {x: x, y: y});
    this.clearHighlightedPos();
    this.gamestate.selPos = null;
    this.draw();
  }

  private help(): void {
    this.gamestate.highlightedPos = this.gameboard.help();
    if(this.gamestate.highlightedPos) {
      this.gamestate.elements[this.gamestate.highlightedPos.e1.y][this.gamestate.highlightedPos.e1.x].rect.isStroked = true;
      this.gamestate.elements[this.gamestate.highlightedPos.e2.y][this.gamestate.highlightedPos.e2.x].rect.isStroked = true;
    }
  }

  private check(): void {
    this.gameboard.check();
    this.clearHighlightedPos();
    this.draw();
  }

  private undo(): void {
    this.gameboard.undo();
    this.clearHighlightedPos();
    this.draw();
  }

  private restart(): void {
    this.gameboard.restart();
    this.clearHighlightedPos();
    this.draw();
  }

  private clearRows(): void {
    this.gameboard.clearRows();
    this.clearHighlightedPos();
    this.draw();
  }

  private draw(): void {
    // Used to store the length of the displayed board before appending
    // Gets updated if rows get deleted
    let lengthT = this.gamestate.elements.length;
    if(this.gameboard.getBoardSizeY() > this.gamestate.elements.length) {
      for(let i = this.gamestate.elements.length; i < this.gameboard.getBoardSizeY(); i++) {
        this.gamestate.elements.push([]);
        for(let j = 0; j < this.gameboard.getBoardSizeX(i); j++) this.appendElement(j, i);
      }
      this.cameras.main.setBounds(0, 0, Constants.width, this.gamestate.elements.length * 30 + Constants.menuHeight);
      this.physics.world.setBounds(0, Constants.height / 2, Constants.width, this.gamestate.elements.length * 30 - Constants.height + Constants.menuHeight);
    }
    if(this.gameboard.getBoardSizeY() < this.gamestate.elements.length) {
      for(let i = this.gameboard.getBoardSizeY(); i < this.gamestate.elements.length; i++) {
        for(let j = 0; j < this.gamestate.elements[i].length; j++) this.deleteElement(j, i);
      }
      for(let i = this.gamestate.elements.length - 1; i >= this.gameboard.getBoardSizeY(); i--) {
        this.gamestate.elements.pop();
      }
      lengthT = this.gamestate.elements.length;
      this.cameras.main.setBounds(0, 0, Constants.width, this.gamestate.elements.length * 30 + Constants.menuHeight);
      this.physics.world.setBounds(0, Constants.height / 2, Constants.width, this.gamestate.elements.length * 30 - Constants.height + Constants.menuHeight);
    }
    if(lengthT == 0) return;
    if(this.gameboard.getBoardSizeX(lengthT - 1) > this.gamestate.elements[lengthT - 1].length) {
      for(let j = this.gamestate.elements[lengthT - 1].length; j < this.gameboard.getBoardSizeX(lengthT - 1); j++) this.appendElement(j, lengthT - 1);
    }
    if(this.gameboard.getBoardSizeX(lengthT - 1) < this.gamestate.elements[lengthT - 1].length) {
      for(let j = this.gameboard.getBoardSizeX(lengthT - 1); j < this.gamestate.elements[lengthT - 1].length; j++) this.deleteElement(j, lengthT - 1);
      for(let j = this.gamestate.elements[lengthT - 1].length - 1; j >= this.gameboard.getBoardSizeX(lengthT - 1); j--) {
        this.gamestate.elements[lengthT - 1].pop();
      }
    }
    for(let i = 0; i < lengthT; i++) {
      for(let j = 0; j < this.gamestate.elements[i].length; j++) {
        if(this.gamestate.elements[i][j].val != this.gameboard.getValue(j, i)) this.updateElement(j, i);
      }
    }
  }

  private createMenu(): void {
    let buttonwidth = Constants.width / 5;
    this.menu.bar = this.add.rectangle(Constants.width / 2, Constants.height - Constants.menuHeight / 2, Constants.width, Constants.menuHeight, 0x000000);
    this.menu.bar.setScrollFactor(0);
    this.menu.bar.depth = 10;

    this.menu.buttons.check.rect = this.add.rectangle(
      buttonwidth / 2 + buttonwidth * 0, 
      Constants.height - Constants.menuHeight / 2, 
      buttonwidth, 
      Constants.menuHeight, 
      0x111111
    );
    this.menu.buttons.check.rect.depth = 11;
    this.menu.buttons.check.rect.strokeColor = 0xffffff;
    this.menu.buttons.check.rect.isStroked = true;
    this.menu.buttons.check.rect.lineWidth = 5;
    this.menu.buttons.undo.rect = this.add.rectangle(
      buttonwidth / 2 + buttonwidth * 1, 
      Constants.height - Constants.menuHeight / 2, 
      buttonwidth, 
      Constants.menuHeight, 
      0x111111
    );
    this.menu.buttons.undo.rect.depth = 11;
    this.menu.buttons.undo.rect.strokeColor = 0xffffff;
    this.menu.buttons.undo.rect.isStroked = true;
    this.menu.buttons.undo.rect.lineWidth = 5;
    this.menu.buttons.help.rect = this.add.rectangle(
      buttonwidth / 2 + buttonwidth * 2, 
      Constants.height - Constants.menuHeight / 2, 
      buttonwidth, 
      Constants.menuHeight, 
      0x111111
    );
    this.menu.buttons.help.rect.depth = 11;
    this.menu.buttons.help.rect.strokeColor = 0xffffff;
    this.menu.buttons.help.rect.isStroked = true;
    this.menu.buttons.help.rect.lineWidth = 5;
    this.menu.buttons.restart.rect = this.add.rectangle(
      buttonwidth / 2 + buttonwidth * 3, 
      Constants.height - Constants.menuHeight / 2, 
      buttonwidth, 
      Constants.menuHeight, 
      0x111111
    );
    this.menu.buttons.restart.rect.depth = 11;
    this.menu.buttons.restart.rect.strokeColor = 0xffffff;
    this.menu.buttons.restart.rect.isStroked = true;
    this.menu.buttons.restart.rect.lineWidth = 5;
    this.menu.buttons.rowClear.rect = this.add.rectangle(
      buttonwidth / 2 + buttonwidth * 4, 
      Constants.height - Constants.menuHeight / 2, 
      buttonwidth, 
      Constants.menuHeight, 
      0x111111
    );
    this.menu.buttons.rowClear.rect.depth = 11;
    this.menu.buttons.rowClear.rect.strokeColor = 0xffffff;
    this.menu.buttons.rowClear.rect.isStroked = true;
    this.menu.buttons.rowClear.rect.lineWidth = 5;

    this.menu.buttons.check.rect.setScrollFactor(0);
    this.menu.buttons.undo.rect.setScrollFactor(0);
    this.menu.buttons.help.rect.setScrollFactor(0);
    this.menu.buttons.restart.rect.setScrollFactor(0);
    this.menu.buttons.rowClear.rect.setScrollFactor(0);

    this.menu.buttons.check.text = this.add.text(
      buttonwidth * 0 + 10, Constants.height - Constants.menuHeight + 10, "Check", { fontFamily: 'monospace', fontSize: '15px'}
    );
    this.menu.buttons.check.text.depth = 12;
    this.menu.buttons.undo.text = this.add.text(
      buttonwidth * 1 + 10, Constants.height - Constants.menuHeight + 10, "Undo", { fontFamily: 'monospace', fontSize: '15px'}
    );
    this.menu.buttons.undo.text.depth = 12;
    this.menu.buttons.help.text = this.add.text(
      buttonwidth * 2 + 10, Constants.height - Constants.menuHeight + 10, "Help", { fontFamily: 'monospace', fontSize: '15px'}
    );
    this.menu.buttons.help.text.depth = 12;
    this.menu.buttons.restart.text = this.add.text(
      buttonwidth * 3 + 10, Constants.height - Constants.menuHeight + 10, "Restart", { fontFamily: 'monospace', fontSize: '15px'}
    );
    this.menu.buttons.restart.text.depth = 12;
    this.menu.buttons.rowClear.text = this.add.text(
      buttonwidth * 4 + 10, Constants.height - Constants.menuHeight + 10, "RowClear", { fontFamily: 'monospace', fontSize: '15px'}
    );
    this.menu.buttons.rowClear.text.depth = 12;

    this.menu.buttons.check.text.setScrollFactor(0);
    this.menu.buttons.undo.text.setScrollFactor(0);
    this.menu.buttons.help.text.setScrollFactor(0);
    this.menu.buttons.restart.text.setScrollFactor(0);
    this.menu.buttons.rowClear.text.setScrollFactor(0);

    this.menu.buttons.check.rect.setInteractive();
    this.menu.buttons.check.rect.on('pointerup', () => this.check());
    this.menu.buttons.undo.rect.setInteractive();
    this.menu.buttons.undo.rect.on('pointerup', () => this.undo());
    this.menu.buttons.help.rect.setInteractive();
    this.menu.buttons.help.rect.on('pointerup', () => this.help());
    this.menu.buttons.restart.rect.setInteractive();
    this.menu.buttons.restart.rect.on('pointerup', () => this.restart());
    this.menu.buttons.rowClear.rect.setInteractive();
    this.menu.buttons.rowClear.rect.on('pointerup', () => this.clearRows());
  }

  private appendElement(x: number, y: number): void {
    let xPos = x * 50 + 25;
    let yPos = y * 30 + 15;
    let properties = this.getProperties(x, y);
    this.gamestate.elements[y].push({
      val: {
        value: this.gameboard.getValue(x, y).value,
        visible: this.gameboard.getValue(x, y).visible,
      },
      rect: this.add.rectangle(xPos, yPos, 50, 30, properties.bgColor),
      number: this.add.text(xPos - 7, yPos - 12 , this.gameboard.getValue(x, y).value.toString(), { fontFamily: 'monospace', color: properties.tColor, fontSize: '25px'}),
    });
    this.gamestate.elements[y][x].rect.depth = 1;
    this.gamestate.elements[y][x].rect.strokeColor = 0xffff00;
    this.gamestate.elements[y][x].rect.lineWidth = 5;
    this.gamestate.elements[y][x].number.depth = 2;
    this.gamestate.elements[y][x].rect.setInteractive();
    this.gamestate.elements[y][x].rect.on('pointerup', () => this.selection(x, y));
  }

  private deleteElement(x: number, y: number): void {
    this.gamestate.elements[y][x].number.destroy();
    this.gamestate.elements[y][x].rect.destroy();
  }

  private updateElement(x: number, y: number): void {
    let properties = this.getProperties(x, y);
    this.gamestate.elements[y][x].rect.fillColor = properties.bgColor;
    this.gamestate.elements[y][x].number.setColor(properties.tColor);
    this.gamestate.elements[y][x].number.setText(this.gameboard.getValue(x, y).value.toString());
    this.gamestate.elements[y][x].val.value = this.gameboard.getValue(x, y).value;
    this.gamestate.elements[y][x].val.visible = this.gameboard.getValue(x, y). visible;
  }

  private clearHighlightedPos(): void {
    if(this.gamestate.highlightedPos) {
      this.gamestate.elements[this.gamestate.highlightedPos.e1.y][this.gamestate.highlightedPos.e1.x].rect.isStroked = false;
      this.gamestate.elements[this.gamestate.highlightedPos.e2.y][this.gamestate.highlightedPos.e2.x].rect.isStroked = false;
      this.gamestate.highlightedPos = null;
    }
  }

  private getProperties(x: number, y: number): {bgColor: number, tColor: string} {
    if(this.gameboard.getValue(x, y).visible) return {bgColor: 0x000000, tColor: '#00ff00'};
    return {bgColor: 0x999999, tColor: '#000000'};
  }
}
